SC17: reworked module file tutorial section

alalazo opened the pull request:
To anybody who would like to review this: :bowing_man: 

Any suggestion related to content, phrasing, spelling, layout, ...., is welcome!

You can view the documentation locally by fetching this PR and:
```console
$ cd lib/spack/docs
$ make
$ <your-favorite-browser> _build/html/index.html
```

##### Modifications
The list of modifications here includes:
- [x] The addition of a brief introductory section on module files
- [x] The addition of a subsection on the overall architecture of module files
- [x] The preparation of a docker image to ease the set-up (thanks @ax3l)
- [x] Reference properly #5582
- [x] Add a section on how to override templates (basically a gentle introduction to `Jinja2`)

I took the freedom of bothering directly a first set of reviewers but really, anybody feel free to comment!

davydden commented:
frankly, i never worked with modules in Spack, but I appreciate that you improve the module tutorial :+1:

alalazo commented:
>  i never worked with modules in Spack, 

@davydden That makes you the perfect reviewer for a tutorial!

davydden commented:
> That makes you the perfect reviewer for a tutorial!

ok, let me see If I can find some time during the week...

davydden commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -1,52 +1,279 @@
+Let's start by summarizing what are module files and how you can use
+them to modify your environment. The idea is to give enough information so that
+people without any previous exposition to them will be able to follow the tutorial
+later on. We'll give also an high-level view of how module files are generated
```
**a** high-level

davydden replied:
I would say "h" is not silent to make it "an", like in "an hour".

davydden commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -1,52 +1,279 @@
+  Lmod
+    Lmod is a module system written in Lua, designed to handle easily hierarchies of
+    module files. It's a drop-in replacement of Environment Modules and works with
+    both of the module file formats generated by Spack.
```
maybe: ...works with both module file formats generated by... ?

davydden commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -1,52 +1,279 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Before  we dive into the hands-on it's worth spending a couple of words to explain how
+module files are generated by Spack. The following diagram provides an high-level view
```
**a** high-level

davydden commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -58,42 +285,78 @@ something like this:
+  ...
+
+but the exact command varies according to your OS. Then we need to prepare
+a ``packages.yaml`` file that instructs Spack to use this external package:
```
maybe be more precise by saying: ... to use externally provided MPI:

mamelara commented:
Checking this out today. Will have my suggestions by COB or tomorrow OOB

alalazo commented:
@mamelara Thanks!

ax3l commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -653,88 +913,124 @@ and regenerating the module files for every package that depends on ``python``:
+  - need constant updates, for instance as soon as a new compiler or MPI library is installed
+
+`Hierarchical module files <http://lmod.readthedocs.io/en/latest/080_hierarchy.html>`_ try to
+overcome these shortcomings by showing at start-up only a a restricted view of what is
```
double `a` :)

ax3l commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -922,63 +1274,255 @@ can add an arbitrary list of virtual providers to the triplet
+  Use "module spider" to find all possible modules.
+  Use "module keyword key1 key2 ..." to search for all possible modules matching any of the "keys".
+
+Both ``MPI`` and ``LAPACK`` providers will now benefict from the same safety features:
```
typo: benefit?

ax3l commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -922,63 +1274,255 @@ can add an arbitrary list of virtual providers to the triplet
+allowed users belong to the same linux group, and access is granted at group level.
+Wouldn't it be nice if people that are not
+yet entitled to use it could receive an helpful message at module load time
+that tells them who to contact in your organization to be inserted in the group?
```
interesting possibilities!
one could also do the same for modules of specific architectures in which an architecture represents a queue of a cluster... :)

alalazo replied:
Sure. Maybe when we have more experience on the use cases we could collect them in some other section of the docs (not the tutorial). The idea here is that with templates you can easily add arbitrary code to module files.

ax3l commented on share/spack/docs/docker/module-file-tutorial/packages.yaml:
```diff
@@ -0,0 +1,9 @@
+  openmpi:
+    buildable: False
+    paths:
+      openmpi@1.10.2: /usr
```
missing newline at eof

mamelara commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -280,20 +527,26 @@ exceptions to the blacklist rules you can use ``whitelist``:
+
 
-you'll see that now the module for ``gcc@6.2.0`` has reappeared:
+you'll see that now the module for ``gcc@7.2.0`` has reappeared:
```
I had difficulty having gcc@7.2.0. Ran the command spack module refresh --module-type tcl -y but that module is still not appearing.

alalazo replied:
Hmm, I went through the tutorial today and everything worked for me (using the Docker image).

mamelara replied:
that might be the case. Like I said some of the steps I had to skip since I wasn't sure what was already set up for me and what wasn't. 

I just wanted to make a comment in case it was a bug or user incompetence

mamelara commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -653,88 +913,124 @@ and regenerating the module files for every package that depends on ``python``:
-  Autoloading python/2.7.12-gcc-6.2.0
-  Autoloading openblas/0.2.19-gcc-6.2.0
-  Autoloading py-numpy/1.11.1-gcc-6.2.0-openblas
+  root@module-file-tutorial:/# module load py-scipy
```
I didn't get this output when I was following the instruction

alalazo replied:
I was also able to follow this. Was your `modules.yaml` the same as the one shown a few lines above?

mamelara replied:
yes, it was. But no output was shown.

alalazo commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -1,52 +1,250 @@
 
 ^^^^^^^^^^^^^^^^^^
-Add a new compiler
+Use a Docker image
```
You're right. I gave it for granted that the purpose of the Docker image was to avoid installing software. It's explicit now. Let me know if the new phrasing is clear enough.

alalazo commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -1,52 +1,250 @@
+
+.. code-block:: console
+
+  $ bin/spack install lmod
```
I rephrased a few sentences, and hopefully now it's clear that the two ways of setting up the environment (use Docker or installing the software manually) are mutually exclusive.

Concerning this command, it worked for me on Ubuntu 14.04 with a particular version of Spack. But I can't possibly follow every variation of OS + flavor of Spack to give a tutorial that permits to "cut and paste" commands and follow along. That's why I advise to use the Docker image.

mamelara commented:
@alalazo  Oh one thing I forgot to mention. Have the image come with an editor. So I would either install nano or use spack to install vim and add it to the environment.

alalazo commented:
> Oh one thing I forgot to mention. Have the image come with an editor. So I would either install nano or use spack to install vim and add it to the environment.

@mamelara It has `emacs`, but I'll add other editors too

mamelara commented:
Oh, haha...I use vim

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -349,7 +631,7 @@ name. We have the possibility to add suffixes to differentiate them:
         -  gcc
       blacklist:
-        -  '%gcc@4.8'
+        -  '%gcc@5.4.0'
```
A couple of lines above this one (but not currently not part of the PR) is the sentence "We have the possibility to add suffixes...".  A better fit for the rest of the tone is "We can add suffixes...".

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -367,19 +649,20 @@ Regenerating module files now we obtain:
 .. code-block:: console
 
-  $ spack module refresh --module-type tcl --delete-tree -y
+  root@module-file-tutorial:/# spack module refresh --module-type tcl --delete-tree -y
```
A couple of lines above this (but not currently part of the PR) is "specify rules that applies", it should be "specify rules that apply".

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -412,26 +695,31 @@ The final result should look like:
   TCL specific directive
     The directives ``naming_scheme`` and ``conflict`` are TCL specific and do not apply
-    to the ``dotkit`` or ``lmod`` sections in the configuration file.
+    to the ``lmod`` section in the configuration file.
```
"and can't be used in the ``lmod`` section of the configuration file."

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -653,88 +944,124 @@ and regenerating the module files for every package that depends on ``python``:
+`Hierarchical module files <http://lmod.readthedocs.io/en/latest/080_hierarchy.html>`_ try to
+overcome these shortcomings by showing at start-up only a restricted view of what is
+available on the system: more specifically only the software that has been installed with
+OS provided compilers. Among this software there will be other - usually more recent - compilers
```
Is it worth calling out that compilers are mutually exclusive?

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -653,88 +944,124 @@ and regenerating the module files for every package that depends on ``python``:
+  #. enable the ``lmod`` file generator
+  #. change the ``tcl`` tag to ``lmod``
+  #. remove ``tcl`` specific directives (``naming_scheme`` and ``conflict``)
+  #. set which compilers are considered ``core_compilers``
```
set -> declare

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -776,113 +1103,169 @@ After modifications the configuration file will be:
-
-The directive ``core_compilers`` accepts a list of compilers; everything built
-using these compilers will create a module in the ``Core`` part of the hierarchy. It is
+The double colon after ``enable`` is intentional and it serves the
```
I liked the fact that this was called out in a note, it's a subtle point and worth highlighting!

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -776,113 +1103,169 @@ After modifications the configuration file will be:
+``Lmod`` took care of swapping the MPI provider for us, and it also substituted the
+``netlib-scalapack`` module to conform to the change in the MPI.
+In this way we can't accidentally pull-in two different MPI providers at the
+same time or load a module file linked to ``openmpi`` when ``mpich`` is also loaded.
```
"or load a module file for a package linked to..."

~~also~~

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -776,113 +1103,169 @@ After modifications the configuration file will be:
+but it still has an asymmetry: ``LAPACK`` providers cover the same semantic role
+as ``MPI`` providers, but yet they are not part of the hierarchy.
+
+To be more practical, this means that despite we have gained an improved consistency in
```
"...means that although we have..."

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -776,113 +1103,169 @@ After modifications the configuration file will be:
+  Currently Loaded Modules:
+    1) gcc/7.2.0   2) openblas/0.2.20   3) openmpi/1.10.2   4) netlib-lapack/3.6.1   5) netlib-scalapack/2.0.2-netlib
+
+Now: having hierarchies that are deeper than ``Core``/``Compiler``/``MPI`` is
```
"Now: having hieararchies..." -> "Hierarchies..." and also "is probably" -> "are probably"

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -776,113 +1103,169 @@ After modifications the configuration file will be:
+Now: having hierarchies that are deeper than ``Core``/``Compiler``/``MPI`` is
+probably still considered "unusual" or "impractical" at many sites, mainly because
+module files are written manually and keeping track of the combinations
+among multiple providers becomes quickly quite involved.
```
"quickly becomes quite"

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -776,113 +1103,169 @@ After modifications the configuration file will be:
+to decide when to show it to the user. The situation becomes more involved as the number of virtual
+dependencies in the hierarchy increases.
+
+Anyhow in Spack we can take advantage of the internal DAG representation that any of
```
"We can take advantage of the DAG that Spack maintains for the installed software  and solve this combinatorial..."

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -776,113 +1103,169 @@ After modifications the configuration file will be:
+
+Anyhow in Spack we can take advantage of the internal DAG representation that any of
+the installed software has, and solve this combinatorial problem in a clean and automated way.
+In some sense deeper hierarchies are feasible due to Spack's ability to manage combinatorial complexity!
```
Spack's ability to manage this combinatorial complexity makes deeper hiearchies feasible.

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -922,63 +1305,255 @@ can add an arbitrary list of virtual providers to the triplet
+Working with templates
+----------------------
+
+As briefly mentioned in the introduction, Spack uses `Jinja2 <http://jinja.pocoo.org/docs/2.9/>`_
```
~~As briefly mentioned in the introduction,~~

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -922,63 +1305,255 @@ can add an arbitrary list of virtual providers to the triplet
+----------------------
+
+As briefly mentioned in the introduction, Spack uses `Jinja2 <http://jinja.pocoo.org/docs/2.9/>`_
+to generate each individual module file. This means that you have at your disposal
```
"This means that you have all of its flexibility and power when it comes to..."

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -922,63 +1305,255 @@ can add an arbitrary list of virtual providers to the triplet
+that will be evaluated and substituted at module generation time.
+The rest of the file is then divided into
+`blocks <http://jinja.pocoo.org/docs/2.9/templates/#template-inheritance>`_
+that can be overridden or extended by users, if need be.
```
~~, if need be~~

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -922,63 +1305,255 @@ can add an arbitrary list of virtual providers to the triplet
+    template_dirs:
+      - $HOME/.spack/templates
+
+This tells Spack to search also another location when looking for template files.
```
"to also search"

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -922,63 +1305,255 @@ can add an arbitrary list of virtual providers to the triplet
+      netlib-scalapack:
+        template: 'group-specific.lua'
+
+If we regenerate once again the module files:
```
"...regenerate the module files one last time:"

hartzell commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -1,52 +1,277 @@
+.. code-block:: console
+
+  $ docker pull alalazo/spack:module_tutorial
+  $ docker run -h module-file-tutorial -it alalazo/spack:module_tutorial
```
Perhaps add `--rm` to this so that the container is automagically cleaned up when it exits?

mamelara commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -103,56 +364,66 @@ Next you should install a few modules that will be used in the tutorial:
+
+Consider now the case that your site has decided that ``CPATH`` and
 ``LIBRARY_PATH`` modifications should not be present in module files. What you can
 do to abide by the rules is to create a configuration file ``~/.spack/modules.yaml``
```
oh boy.... 😛

alalazo commented:
@mamelara I just updated the image, which now contains vim. Can you check if the PR now is good to go in your opinion?

@mamelara @hartzell Thank you very much for the feedback!

alalazo commented:
@tgamblin ping

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -1,52 +1,278 @@
-Prerequisites
--------------
+-----------------------
+Set up for the tutorial
```
_Setup_ for the tutorial

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -1,52 +1,278 @@
+If you don't have Docker installed or for any other reason you
+prefer to work locally, follow instead :ref:`module_file_tutorial_work_locally`
+to know how to clone Spack and install the software.
+Be aware that the set-up will take longer and that the details of the snippets
```
I think "longer" is an understatement. The last time I tried to build the full set of software in the tutorial with Spack-alone, it took 3 hours. I see that you're offloading some of this on the system package manager. Many users might not even have access to that. I'm not sure if I can think of a better solution than telling them to use Docker.

alalazo replied:
> Many users might not even have access to that. 

The manual setup must be taken as a guideline, not as a set of cut and paste commands. There are  far too many possible options to cover if we go that way. For instance, users on a cluster  won't have access to a system package manager, but they probably have already an external MPI installed and can use that in `packages.yaml`.

alalazo replied:
> The last time I tried to build the full set of software in the tutorial with Spack-alone, it took 3 hours

Confirmed: almost 3 hours to rebuild the image on my laptop.

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -1,52 +1,278 @@
+^^^^^^^^^^^^^^^^^^^^^^
+
+If you don't feel like using a container, you can set-up your environment
+locally. Let's start by cloning Spack repository and moving to the directory
```
cloning _the_ Spack repository

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -1,52 +1,278 @@
+
+.. code-block:: console
+
+  $ git clone https://github.com/LLNL/spack.git
```
spack/spack

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -1,52 +1,278 @@
+  $ git clone https://github.com/LLNL/spack.git
+  $ cd spack
+
+From here we'll be building up the required stack of software.
```
building the

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -1,52 +1,278 @@
+
+  $ bin/spack install lmod
+  ...
+  output omitted
```
Probably don't need this line or the dots

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -1,52 +1,278 @@
+
+.. code-block:: console
+
+  $ . $(spack location -i lmod)/lmod/lmod/init/bash
```
I think there is a `profile` script in the same directory that should be more generic. Just in case we have any zsh users 😉

alalazo replied:
I hope at SC17 we'll be using the docker image :smile:

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -1,52 +1,278 @@
+""""""""""""""""""
+
+The second step is to build a recent compiler. On first use, Spack
+scans the environment and picks up automatically the
```
replace "picks up automatically" with "automatically picks up" or "automatically locates"

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -58,42 +284,78 @@ something like this:
-build a limited set of packages with multiple compilers. If you do not already
-have multiple compilers listed by ``spack compilers``, you should build one
-with Spack:
+Let's bootstrap a more recent compiler with the one that was detected automatically:
```
"automatically detected"

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -58,42 +284,78 @@ something like this:
+
+  $ spack install gcc@7.2.0
+  ...
+  Wait a long time
```
But like, seriously. Takes an hour on a laptop. Not feasible for an in-person tutorial, but is nice for an "on your own time" tutorial

alalazo replied:
I wrote this tutorial thinking at a stand-alone piece of documentation, unrelated to the context of SC. The idea is that we can reuse this for an in-person tutorial, not *use this just for it*.

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -103,56 +365,66 @@ Next you should install a few modules that will be used in the tutorial:
 .. code-block:: console
 
-  $ module avail
+  root@module-file-tutorial:/# module av
```
I think it's best to avoid shortcuts and abbreviations. I like `module avail` better.

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -103,56 +365,66 @@ Next you should install a few modules that will be used in the tutorial:
 .. code-block:: console
 
-  $ module show gcc-6.2.0-gcc-4.8-twd5nqg
+  root@module-file-tutorial:/# module show gcc-7.2.0-gcc-5.4.0-go3z4hb
```
The bash prompt will look different for every user, right? I think it's better to stay consistent with the rest of Spack's documentation and use `$` everywhere.

alalazo replied:
Nope. It's a docker image. You'll see exactly that.

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -226,41 +503,41 @@ and regenerate the module files:
 
-If you pay careful attention you'll see though that we went too far in blacklisting modules:
-the module for ``gcc@6.2.0`` disappeared as it was bootstrapped with ``gcc@4.8``. To specify
+If look closely you'll see though that we went too far in blacklisting modules:
```
If _you_ look closely

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -653,88 +945,124 @@ and regenerating the module files for every package that depends on ``python``:
+
+So far we worked with non-hierarchical module files, i.e. with module files
+that are all generated in the same root directory and don't attempt to modify
+dynamically the ``MODULEPATH``. This results in a flat module structure where
```
dynamically modify

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -653,88 +945,124 @@ and regenerating the module files for every package that depends on ``python``:
+  Use "module keyword key1 key2 ..." to search for all possible modules matching any of the "keys".
+
+This layout is quite simple to deploy, but you can see from the above snippet
+that nothing prevents users to load incompatible sets of modules:
```
prevents users _from_ loading

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -653,88 +945,124 @@ and regenerating the module files for every package that depends on ``python``:
+OS provided compilers. Among this software there will be other - usually more recent - compilers
+that, once loaded, will prepend new directories to ``MODULEPATH`` unlocking all the software
+that was compiled with them. This "unlocking" idea can then be extended arbitrarily to
+virtual dependencies, as we'll see in the following.
```
following _section_

adamjstewart commented on lib/spack/docs/tutorial_modules.rst:
```diff
@@ -922,63 +1306,255 @@ can add an arbitrary list of virtual providers to the triplet
+Let's assume one of our software is protected by group membership:
+allowed users belong to the same linux group, and access is granted at group level.
+Wouldn't it be nice if people that are not
+yet entitled to use it could receive an helpful message at module load time
```
_a_ helpful message

alalazo commented:
@tgamblin For the points you mentioned above:

1. I'll definitely look into that
2. I added a few lines to explain what vendored means, let me know if you want more / different things
3. The images were already using `ubuntu:16.04`. I updated the repository in the Docker image to use the latest `spack/spack/develop`. May be worth point it to a tag if we release before SC?

Should we move the `Dockerfile` + configuration in another repository of the Spack organization before SC17? Or do we want to think better about it and do it later?

It would be cool if we could automatically deploy to Dockerhub via some CI build, at some point...

@adamjstewart Thanks for the review! Really appreciate you took the time to go through the tutorial :bowing_man:

adamjstewart commented:
Looking at the other tutorials, I don't see anywhere where we explain how to add custom environment variables to be added to the module files by adding `setup_environment` in the `package.py`. I can't tell if it would fit better in the Modules tutorial or the Advanced Packaging tutorial, but I'm leaning towards Modules
